File: recursion/tower_of_hanoi.cpp
/**
 * @file
 * @brief Tower of Hanoi Implementation
 */

#include <iostream>

/**
 * @brief Moves the disks
 * @param fp From pole
 * @param tp To pole
 */
void moveDisk(char fp, char tp) {
  std::cout << "moving disk from " << fp << " to " << tp << std::endl;
}

/**
 * @brief Recursive function to move tower
 * @param height Height of the tower
 * @param fromPole Source pole
 * @param toPole Destination pole
 * @param withPole Auxiliary pole
 */
void moveTower(int height, char fromPole, char toPole, char withPole) {
  if (height >= 1) {
    moveTower(height - 1, fromPole, withPole, toPole);
    moveDisk(fromPole, toPole);
    moveTower(height - 1, withPole, toPole, fromPole);
  }
}

int main() {
  int height = 3;
  moveTower(height, 'A', 'B', 'C');
  return 0;
}

File: recursion/tower_of_hanoi.rs
/**
 * Tower of Hanoi Implementation in Rust
 */

fn move_disk(fp: char, tp: char) {
    println!("moving disk from {} to {}", fp, tp);
}

fn move_tower(height: i32, from_pole: char, to_pole: char, with_pole: char) {
    if height >= 1 {
        move_tower(height - 1, from_pole, with_pole, to_pole);
        move_disk(from_pole, to_pole);
        move_tower(height - 1, with_pole, to_pole, from_pole);
    }
}

fn main() {
    let height = 3;
    move_tower(height, 'A', 'B', 'C');
}
